solidity/contracts/token/extensions/TokenBridgeCctp.sol

```diff
@@ -0,0 +1,138 @@
+     */
+    event DomainAdded(uint32 indexed hyperlaneDomain, uint32 circleDomain);
+
+    /**
+     * @notice Raised when the version in use by the TokenMessenger
+     * is not recognized
+     */
+    error UnsupportedCCTPVersion(uint32 wrongVersion);
+
+    constructor(
+        address _erc20,
+        uint256 _scale,
+        address _mailbox,
+        ITokenMessenger _tokenMessenger
+    ) HypERC20Collateral(_erc20, _scale, _mailbox) {
+        tokenMessenger = _tokenMessenger;
+    }
+
+    /**
+     * @notice Adds a new mapping between a Hyperlane domain and a Circle domain.
+     * @param _hyperlaneDomain The Hyperlane domain.
+     * @param _circleDomain The Circle domain.
+     */
+    function addDomain(
+        uint32 _hyperlaneDomain,
+        uint32 _circleDomain
+    ) external onlyOwner {
+        hyperlaneDomainToCircleDomain[_hyperlaneDomain] = _circleDomain;
+
+        emit DomainAdded(_hyperlaneDomain, _circleDomain);
+    }
+
+    function transferRemote(
+        uint32 _destination,
+        bytes32 _recipient,
+        uint256 _amount
+    )
+        external
+        payable
+        override(ITokenBridge, TokenRouter)
+        returns (bytes32 messageId)
+    {
+        messageId = TokenRouter._transferRemote(
+            _destination,
+            _recipient,
+            _amount,
+            msg.value
+        );
+
+        // Has to be called after _transferRemote
+        // in order for _transferFromSender to be
+        // executed first
+        _transferToCctpTokenMessenger(_destination, _amount);
+    }
+
+    function quoteTransferRemote(
+        uint32 _destination,
+        bytes32 /* _recipient */,
+        uint256 /* _amount */
+    ) external view override returns (Quote[] memory quotes) {
+        quotes = new Quote[](1);
+        quotes[0] = Quote(address(0), _quoteGasPayment(_destination));
+    }
+
+    function _transferToCctpTokenMessenger(
+        uint32 _destination,
+        uint256 _amount
+    ) internal {
+        wrappedToken.approve(address(tokenMessenger), _amount);
+        uint32 circleDomain = hyperlaneDomainToCircleDomain[_destination];
+        uint32 version = tokenMessenger.messageBodyVersion();
+
+        bytes32 router = _mustHaveRemoteRouter(_destination);
+
+        if (version == CCTP_VERSION_1) {
+            tokenMessenger.depositForBurn(
+                _amount,
+                circleDomain,
+                router,
+                address(wrappedToken)
+            );
+        } else if (version == CCTP_VERSION_2) {
+            ITokenMessengerV2(address(tokenMessenger)).depositForBurn(
+                _amount,
+                circleDomain,
+                router,
+                address(wrappedToken),
+                bytes32(0),
+                CCTP_V2_DEFAULT_MAX_FEE,
+                CCTP_V2_DEFAULT_MIN_FINALITY_THRESHOLD
+            );
+        } else {
+            revert UnsupportedCCTPVersion(version);
+        }
```

Comment:
tokenMessenger is immutable so each deployment of this will only ever use 1 branch
in this case it makes sense to have 2 separate implementations that assert the version matches at deployment time and does not need to have branching logic here

Reply:
Makes sense, I'll prob be able to do it in the coming days, but can't test it until I'm back